### PR TITLE
[JENKINS-29776] Add Job DSL support for Promoted Builds Plugin

### PR DESCRIPTION
**Pull request by denschu**
The DSL Extension needs to reload the Job configuration after adding a Promotion, so I moved to
Jenkins 1.560 which includes this feature (doReload). This change
includes also some FindBugs fixes.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -4,17 +4,17 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.554.2</version>
+    <version>1.560</version>
```
Comment (oleg-nenashev) [inline pom.xml]: Better to bump the dependency to the LTS release (e.g. 1.565)

File: pom.xml [hunk 2]
```diff
@@ -4,17 +4,17 @@
...
   </parent>
 
   <artifactId>promoted-builds</artifactId>
-  <version>2.24.2-SNAPSHOT</version>
+  <version>2.25-SNAPSHOT</version>
```
Comment (oleg-nenashev) [inline pom.xml]: No need to change it in the PR. It will cause merge conflicts if there is a hotfix going ahead your PR.

File: pom.xml [hunk 3]
```diff
@@ -145,4 +158,4 @@
       </plugin>
     </plugins>
   </build>
-</project>
```
Comment (oleg-nenashev) [inline pom.xml]: Please avoid unrelated changes. They decrease the PR readability and also may cause merge conflicts

File: src/main/java/hudson/plugins/promoted_builds/JobPropertyImpl.java [hunk 4]
```diff
@@ -79,6 +79,13 @@ public JobPropertyImpl(AbstractProject<?,?> owner) throws Descriptor.FormExcepti
         init();
     }
 
+    /**
+     * Programmatic construction.
+     */
+    public JobPropertyImpl(Set<String> activeProcessNames) throws Descriptor.FormException, IOException {
+        this.activeProcessNames.addAll(activeProcessNames);
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/JobPropertyImpl.java]: :bug: Public constructor without property owner setup

File: src/main/java/hudson/plugins/promoted_builds/Promotion.java [hunk 5]
```diff
@@ -316,10 +320,13 @@ protected void post2(BuildListener listener) throws Exception {
                 }
 
                 // tickle PromotionTriggers
-                for (AbstractProject<?,?> p : Jenkins.getInstance().getAllItems(AbstractProject.class)) {
-                    PromotionTrigger pt = p.getTrigger(PromotionTrigger.class);
-                    if (pt!=null)
-                        pt.consider(Promotion.this);
+                Jenkins jenkins = Jenkins.getInstance();
+                if(jenkins != null){
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/Promotion.java]: This update is reasonable, but IMO it would be better to polish the code in a separate pull request. Does it really happens in JobDSL?

File: src/main/java/hudson/plugins/promoted_builds/PromotionProcess.java [hunk 6]
```diff
@@ -498,7 +498,11 @@ public String getId() {
     }
 
     public DescriptorImpl getDescriptor() {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/PromotionProcess.java]: Descriptor should be never null. Better to throw exception here

**Comment by oleg-nenashev** [comment]
@denschu 
Why did you need massive checks of <code>Jenkins.getInstance()</code> results? It's a known issue in many plugins, but static analysis warnings are not the blocker. Do you need this fix in this plugin to mke Job DSL working? I'm not sure.

If you want to modify methods, the best way is to create a new Util method, which would be a copy of https://github.com/jenkinsci/jenkins/blob/0a5b03688ba6884429bb3d2a06be8b7955694420/core/src/main/java/jenkins/model/Jenkins.java#L689-L702 . Such check is enough in the most of cases. E.g. field validation always requires Jenkins to be started up => no need to implement new logic.

File: src/main/java/hudson/plugins/promoted_builds/jobdsl/ConditionsContext.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
+package hudson.plugins.promoted_builds.jobdsl;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/jobdsl/ConditionsContext.java]: maybe promoted_builds.integrations.jobdsl
There will be definitely an entry for Pipeline Plugin (fka Workflow) at some point

File: src/main/java/hudson/plugins/promoted_builds/jobdsl/ConditionsContext.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import javaposse.jobdsl.dsl.Context;
+import javaposse.jobdsl.dsl.FileJobManagement;
+import javaposse.jobdsl.dsl.helpers.BuildParametersContext;
+
+import org.apache.commons.io.FileUtils;
+
+public class ConditionsContext implements Context {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/jobdsl/ConditionsContext.java]: Such code duplication increases maintenance efforts

**Comment by denschu** [comment]
Hi @oleg-nenashev,
thank you for the review :-) The reason for the checks was that after the update to 1.560 the build was not successful, because Findbugs complained about it. I can reimplement it with a Util method. This sounds good :+1: Should I create a separate PR for the update to 1.565?

File: pom.xml (hunk 1 above)
Comment (denschu) [inline pom.xml]: Ok, I will fix that.

File: pom.xml (hunk 2 above)
Comment (denschu) [inline pom.xml]: Ok :-)

File: pom.xml (hunk 3 above)
Comment (denschu) [inline pom.xml]: I'll remove that glitch.

File: src/main/java/hudson/plugins/promoted_builds/JobPropertyImpl.java (hunk 4 above)
Comment (denschu) [inline src/main/java/hudson/plugins/promoted_builds/JobPropertyImpl.java]: This change is needed to reuse the class for the DSL plugin. Otherwise I would have to create a complete new class like JobDslPromotionProcess. But I tried to avoid that as often as possible to reduce maintenance overhead.

File: src/main/java/hudson/plugins/promoted_builds/Promotion.java (hunk 5 above)
Comment (denschu) [inline src/main/java/hudson/plugins/promoted_builds/Promotion.java]: Yep, I'll move that to a separate PR.

**Comment by oleg-nenashev** [comment]
I've just started the review, still need to dig into the implementation details. Seems the PR creates copies of much classes, but it would be great to avoid it in order to reduce mantenance efforts.

> Should I create a separate PR for the update to 1.565?

No, it's better to do it within this pull-request.

**Comment by denschu** [comment]
Hi @oleg-nenashev, 
I've prepared a PR which doesn't contain the checks and fixes your mentioned issues. Just let me know if I should push it into this PR here or create a new one. 

One info concerning code duplication: The *Context-classes (e.g. ConditionsContext) are necessary for the Job DSL. The classes JobDslPromotionProcess and JobDslManualCondition are responsible for holding the created nodes from the default Job DSL core components e.g. BuildParametersContext and StepContext, which we are reusing here. 

Thanks in advance
Dennis

**Comment by denschu** [comment]
Hi @oleg-nenashev,
I have created a new PR (https://github.com/jenkinsci/promoted-builds-plugin/pull/82) which includes only the changes for the Job DSL and not the Findbugs stuff. I hope that is now easier for you to understand what I've modified.

**Comment by fgauthier-boutin** [comment]
FYI,
I'm temporarily using the [job-dsl-promotions plugin](https://github.com/codecentric/job-dsl-promotions-plugin) patiently waiting for this request to be finally integrated.

This morning, I updated Job DSL to version 1.43 (from 1.42) and now I get this error when generating a manual criteria:
`ERROR: (ConditionsContext.groovy, line 64) Could not find matching constructor for: javaposse.jobdsl.dsl.helpers.BuildParametersContext()`

Example, [here](https://github.com/codecentric/job-dsl-promotions-plugin/blob/5cdac32b0ab3c9494fe3251642f0efb7ad31909d/src/test/resources/example-dsl.groovy).

```
manual('tester'){
      parameters{
          textParam("parameterName","defaultValue","description")
      }
  }
```
I tried different combinations without luck.

I'm pretty sure the problem is [this line of code found here](https://github.com/codecentric/job-dsl-promotions-plugin/blob/ddfdfcc21c2e76ed522da0b2d31f4b6a83aae093/src/main/java/org/jenkinsci/plugins/jobdsl/promotions/ConditionsContext.java#L55):
`parameters(parametersClosure);`


BTW, please, don't forget to support the "Keep build forever" action which is part of the Promotion Plugin as well.

Thank you!

**Comment by Jacek-Sniecikowski-A1** [comment]
Any update on this? I'd love to be able to use Build Promotion in JobDSL.

**Comment by oleg-nenashev** [comment]
@Jacek-Sniecikowski-A1 
This PR is  closed. See https://github.com/jenkinsci/promoted-builds-plugin/pull/82
